kernel/src/process_checker.rs

```diff
@@ -0,0 +1,133 @@
+    fn require_credentials(&self) -> bool;
+    fn check_credentials(
+        &self,
+        credentials: TbfFooterV2Credentials,
+        binary: &'a [u8],
+    ) -> Result<(), (ErrorCode, TbfFooterV2Credentials, &'a [u8])>;
+}
+
+/// Default implementation.
+impl<'a> AppCredentialsChecker<'a> for () {
+    fn set_client(&self, _client: &'a dyn Client<'a>) {}
+    fn require_credentials(&self) -> bool {
+        false
+    }
+
+    fn check_credentials(
+        &self,
+        credentials: TbfFooterV2Credentials,
+        binary: &'a [u8],
+    ) -> Result<(), (ErrorCode, TbfFooterV2Credentials, &'a [u8])> {
+        Err((ErrorCode::NOSUPPORT, credentials, binary))
+    }
+}
+
+/// Whether two processes have the same Application Identifier; two
+/// processes with the same Application Identifier cannot run concurrently.
+pub trait AppUniqueness {
+    /// Returns whether `process_a` and `process_b` have a different identifier,
+    /// and so can run concurrently. If this returns `false`, the kernel
+    /// will not run `process_a` and `process_b` at the same time.
+    fn different_identifier(&self, _process_a: &dyn Process, _process_b: &dyn Process) -> bool;
+
+    /// Return whether there is a currently running process that has
+    /// the same application identifier as `process` OR the same short
+    /// ID as `process`. This means that if `process` is currently
+    /// running, `has_unique_identifier` returns false.
+    fn has_unique_identifiers(
+        &self,
+        process: &dyn Process,
+        processes: &[Option<&dyn Process>],
+    ) -> bool {
+        let len = processes.len();
+        // If the process is running or not runnable it does not have
+        // a unique identifier; these two states describe a process
+        // that is potentially runnable, dependent on checking for
+        // identifier uniqueness at runtime.
+        if process.get_state() != State::CredentialsApproved
+            && process.get_state() != State::Terminated
+        {
+            return false;
+        }
+
+        // Note that this causes `process` to compare against itself;
+        // however, since `process` should not be running, it will
+        // not check the identifiers and say they are different. This means
+        // this method returns false if the process is running.
+        for i in 0..len {
+            let checked_process = processes[i];
+            let diff = checked_process.map_or(true, |other| {
+                !other.is_running()
+                    || (self.different_identifier(process, other)
+                        && other.short_app_id() != process.short_app_id())
+            });
+            if !diff {
+                return false;
+            }
+        }
+        true
+    }
+}
+
+/// Default implementation.
+impl AppUniqueness for () {
+    fn different_identifier(&self, _process_a: &dyn Process, _process_b: &dyn Process) -> bool {
+        true
```

Comment:
What is the value of this default implementation? Given that the correct operation of has_unique_identifiers is a security concern, it seems strange to have a default implementation of AppUniqueness for which the same process would be allowed to run twice (since passing the same process for process_a and process_b here would still return true)

Reply:
Note that a process can return `true` when compared with itself if it has a locally unique identifier. The default implementation gives all processes a locally unique identifier so it always returns true. To test whether it's a reference to the same process structure should occur outside this method call; this is just an identifier check.